// vite-react-ssg/dist/client/single-page.mjs
import { createRoot, hydrateRoot } from 'react-dom/client';
import { HelmetProvider } from 'react-helmet-async';
import React from 'react';
import { d as documentReady } from '../shared/vite-react-ssg.a292c181.mjs';
export { C as ClientOnly, H as Head } from '../shared/vite-react-ssg.a292c181.mjs';
import { d as deserializeState } from '../shared/vite-react-ssg.579feabb.mjs';

function ViteReactSSG(App, fn, options = {}) {
  const {
    transformState,
    rootContainer = "#root",
    ssrWhenDev,
    getStyleCollector = null
  } = options;
  if (process.env.NODE_ENV === "development" && ssrWhenDev !== void 0)
    console.warn("[vite-react-ssg] `ssrWhenDev` option is no longer needed. If you want to use csr, just replace `vite-react-ssg dev` with `vite`.");
  const isClient = typeof window !== "undefined";
  async function createRoot$1(client = false, routePath) {
    const appRenderCallbacks = [];
    const onSSRAppRendered = client ? () => {
    } : (cb) => appRenderCallbacks.push(cb);
    const triggerOnSSRAppRendered = () => {
      return Promise.all(appRenderCallbacks.map((cb) => cb()));
    };
    const context = {
      isClient,
      onSSRAppRendered,
      triggerOnSSRAppRendered,
      initialState: {},
      transformState,
      routePath,
      getStyleCollector,
      routes: void 0,
      routerOptions: void 0,
      base: "/",
      app: App
    };
    if (client) {
      await documentReady();
      context.initialState = transformState?.(window.__INITIAL_STATE__ || {}) || deserializeState(window.__INITIAL_STATE__);
    }
    await fn?.(context);
    const initialState = context.initialState;
    return {
      ...context,
      initialState
    };
  }
  if (isClient) {
    (async () => {
      const container = typeof rootContainer === "string" ? document.querySelector(rootContainer) : rootContainer;
      if (!container) {
        if (typeof $jsdom === "undefined")
          console.warn("[vite-react-ssg] Root container not found.");
        return;
      }
      await createRoot$1(true);
      const app = /* @__PURE__ */ React.createElement(HelmetProvider, null, App);
      const isSSR = document.querySelector("[data-server-rendered=true]") !== null;
      if (!isSSR && process.env.NODE_ENV === "development") {
        const root = createRoot(container);
        React.startTransition(() => {
          root.render(app);
        });
      } else {
        React.startTransition(() => {
          hydrateRoot(container, app);
        });
      }
    })();
  }
  return createRoot$1;
}

export { ViteReactSSG };
