// vite-react-ssg/dist/client/single-page.cjs
'use strict';

const client = require('react-dom/client');
const reactHelmetAsync = require('react-helmet-async');
const React = require('react');
const ClientOnly = require('../shared/vite-react-ssg.c1d49976.cjs');
const state = require('../shared/vite-react-ssg.bee8a5a9.cjs');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const React__default = /*#__PURE__*/_interopDefaultCompat(React);

function ViteReactSSG(App, fn, options = {}) {
  const {
    transformState,
    rootContainer = "#root",
    ssrWhenDev,
    getStyleCollector = null
  } = options;
  if (process.env.NODE_ENV === "development" && ssrWhenDev !== void 0)
    console.warn("[vite-react-ssg] `ssrWhenDev` option is no longer needed. If you want to use csr, just replace `vite-react-ssg dev` with `vite`.");
  const isClient = typeof window !== "undefined";
  async function createRoot(client = false, routePath) {
    const appRenderCallbacks = [];
    const onSSRAppRendered = client ? () => {
    } : (cb) => appRenderCallbacks.push(cb);
    const triggerOnSSRAppRendered = () => {
      return Promise.all(appRenderCallbacks.map((cb) => cb()));
    };
    const context = {
      isClient,
      onSSRAppRendered,
      triggerOnSSRAppRendered,
      initialState: {},
      transformState,
      routePath,
      getStyleCollector,
      routes: void 0,
      routerOptions: void 0,
      base: "/",
      app: App
    };
    if (client) {
      await ClientOnly.documentReady();
      context.initialState = transformState?.(window.__INITIAL_STATE__ || {}) || state.deserializeState(window.__INITIAL_STATE__);
    }
    await fn?.(context);
    const initialState = context.initialState;
    return {
      ...context,
      initialState
    };
  }
  if (isClient) {
    (async () => {
      const container = typeof rootContainer === "string" ? document.querySelector(rootContainer) : rootContainer;
      if (!container) {
        if (typeof $jsdom === "undefined")
          console.warn("[vite-react-ssg] Root container not found.");
        return;
      }
      await createRoot(true);
      const app = /* @__PURE__ */ React__default.createElement(reactHelmetAsync.HelmetProvider, null, App);
      const isSSR = document.querySelector("[data-server-rendered=true]") !== null;
      if (!isSSR && process.env.NODE_ENV === "development") {
        const root = client.createRoot(container);
        React__default.startTransition(() => {
          root.render(app);
        });
      } else {
        React__default.startTransition(() => {
          client.hydrateRoot(container, app);
        });
      }
    })();
  }
  return createRoot;
}

exports.ClientOnly = ClientOnly.ClientOnly;
exports.Head = ClientOnly.Head;
exports.ViteReactSSG = ViteReactSSG;
